File: content/develop/use-cases/time-series-dashboard/dotnet/SensorSimulator.cs
using System.Globalization;

namespace TimeSeriesDashboardDemo;

public sealed record SensorDefinition(
    string SensorId,
    string Zone,
    string Unit,
    double MinValue,
    double MaxValue,
    double Baseline,
    double Drift,
    double SpikeChance,
    double SpikeSize)
{
    public string SensorType => "power_consumption";

    public string Key => $"ts:sensor:{SensorType}:{SensorId}";

    public IReadOnlyDictionary<string, string> Labels =>
        new Dictionary<string, string>
        {
            ["site"] = "demo",
            ["sensor_type"] = SensorType,
            ["sensor_id"] = SensorId,
            ["zone"] = Zone,
            ["unit"] = Unit
        };
}

public sealed record SensorSample(SensorDefinition Sensor, double Value);

public static class SensorCatalog
{
    public static IReadOnlyList<SensorDefinition> Sensors { get; } =
        new[]
        {
            new SensorDefinition(
                SensorId: "power-1",
                Zone: "north",
                Unit: "watts",
                MinValue: 320.0,
                MaxValue: 920.0,
                Baseline: 480.0,
                Drift: 18.0,
                SpikeChance: 0.10,
                SpikeSize: 120.0),
            new SensorDefinition(
                SensorId: "power-2",
                Zone: "central",
                Unit: "watts",
                MinValue: 320.0,
                MaxValue: 920.0,
                Baseline: 560.0,
                Drift: 20.0,
                SpikeChance: 0.08,
                SpikeSize: 140.0),
            new SensorDefinition(
                SensorId: "power-3",
                Zone: "south",
                Unit: "watts",
                MinValue: 320.0,
                MaxValue: 920.0,
                Baseline: 640.0,
                Drift: 24.0,
                SpikeChance: 0.12,
                SpikeSize: 160.0)
        };
}

public sealed class SensorSimulator
{
    private readonly IReadOnlyList<SensorDefinition> _sensors;
    private readonly Dictionary<string, double> _values;
    private readonly Random _random = new();

    public SensorSimulator(IReadOnlyList<SensorDefinition> sensors)
    {
        _sensors = sensors;
        _values = sensors.ToDictionary(sensor => sensor.SensorId, sensor => sensor.Baseline);
    }

    public IReadOnlyList<SensorSample> NextSamples()
    {
        var samples = new List<SensorSample>(_sensors.Count);

        foreach (var sensor in _sensors)
        {
            var current = _values[sensor.SensorId];
            var drift = RandomBetween(-sensor.Drift, sensor.Drift);
            var pull = (sensor.Baseline - current) * 0.12;
            var value = current + drift + pull;

            if (sensor.SpikeChance > 0 && _random.NextDouble() < sensor.SpikeChance)
            {
                value += RandomBetween(0.5, 1.0) * sensor.SpikeSize;
            }

            value = Math.Clamp(value, sensor.MinValue, sensor.MaxValue);
            value = Math.Round(value, 2, MidpointRounding.AwayFromZero);

            _values[sensor.SensorId] = value;
            samples.Add(new SensorSample(sensor, value));
        }

        return samples;
    }

    private double RandomBetween(double min, double max) => min + (_random.NextDouble() * (max - min));
}

File: content/develop/use-cases/time-series-dashboard/go/sensor_simulator.go
package main

import (
	"math"
	"math/rand"
)

type SensorDefinition struct {
	SensorID    string
	Zone        string
	Unit        string
	MinValue    float64
	MaxValue    float64
	Baseline    float64
	Drift       float64
	SpikeChance float64
	SpikeSize   float64
}

func (s SensorDefinition) SensorType() string {
	return "power_consumption"
}

func (s SensorDefinition) Key() string {
	return "ts:sensor:" + s.SensorType() + ":" + s.SensorID
}

func (s SensorDefinition) Labels() map[string]string {
	return map[string]string{
		"site":        "demo",
		"sensor_type": s.SensorType(),
		"sensor_id":   s.SensorID,
		"zone":        s.Zone,
		"unit":        s.Unit,
	}
}

type SensorSample struct {
	Sensor SensorDefinition
	Value  float64
}

var Sensors = []SensorDefinition{
	{
		SensorID:    "power-1",
		Zone:        "north",
		Unit:        "watts",
		MinValue:    320.0,
		MaxValue:    920.0,
		Baseline:    480.0,
		Drift:       18.0,
		SpikeChance: 0.10,
		SpikeSize:   120.0,
	},
	{
		SensorID:    "power-2",
		Zone:        "central",
		Unit:        "watts",
		MinValue:    320.0,
		MaxValue:    920.0,
		Baseline:    560.0,
		Drift:       20.0,
		SpikeChance: 0.08,
		SpikeSize:   140.0,
	},
	{
		SensorID:    "power-3",
		Zone:        "south",
		Unit:        "watts",
		MinValue:    320.0,
		MaxValue:    920.0,
		Baseline:    640.0,
		Drift:       24.0,
		SpikeChance: 0.12,
		SpikeSize:   160.0,
	},
}

type SensorSimulator struct {
	sensors []SensorDefinition
	values  map[string]float64
}

func NewSensorSimulator(sensors []SensorDefinition) *SensorSimulator {
	values := make(map[string]float64, len(sensors))
	for _, sensor := range sensors {
		values[sensor.SensorID] = sensor.Baseline
	}
	return &SensorSimulator{
		sensors: sensors,
		values:  values,
	}
}

func (s *SensorSimulator) NextSamples() []SensorSample {
	samples := make([]SensorSample, 0, len(s.sensors))
	for _, sensor := range s.sensors {
		current := s.values[sensor.SensorID]
		drift := randomBetween(-sensor.Drift, sensor.Drift)
		pull := (sensor.Baseline - current) * 0.12
		value := current + drift + pull

		if sensor.SpikeChance > 0 && rand.Float64() < sensor.SpikeChance {
			value += randomBetween(0.5, 1.0) * sensor.SpikeSize
		}

		value = math.Max(sensor.MinValue, math.Min(sensor.MaxValue, value))
		value = math.Round(value*100) / 100

		s.values[sensor.SensorID] = value
		samples = append(samples, SensorSample{
			Sensor: sensor,
			Value:  value,
		})
	}

	return samples
}

func randomBetween(min, max float64) float64 {
	return min + rand.Float64()*(max-min)
}
